### ISSUE TITLE
error in version 0.5.2: Channel closed before HTTP/2 preface completed.

### ISSUE THREAD

**Issue by mobilesoftsmith** [relayrides/pushy#258]
I am suing 0.5.2. I can send notification most time. But sometimes I got this error:

```
[nioEventLoopGroup-2-1] INFO com.relayrides.pushy.apns.ApnsClient - Failed to connect.
java.lang.IllegalStateException: Channel closed before HTTP/2 preface completed.
	at com.relayrides.pushy.apns.ApnsClient$2.operationComplete(ApnsClient.java:495)
	at com.relayrides.pushy.apns.ApnsClient$2.operationComplete(ApnsClient.java:489)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:683)
	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:568)
	at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:407)
	at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
	at io.netty.channel.AbstractChannel$CloseFuture.setClosed(AbstractChannel.java:1042)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.doClose0(AbstractChannel.java:679)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:657)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:601)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.close(DefaultChannelPipeline.java:1416)
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeCloseNow(ChannelHandlerInvokerUtil.java:133)
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeClose(DefaultChannelHandlerInvoker.java:277)
	at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:226)
	at io.netty.handler.ssl.SslHandler$8.operationComplete(SslHandler.java:1432)
	at io.netty.handler.ssl.SslHandler$8.operationComplete(SslHandler.java:1418)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:683)
	at io.netty.util.concurrent.DefaultPromise.notifyLateListener(DefaultPromise.java:624)
	at io.netty.util.concurrent.DefaultPromise.addListener(DefaultPromise.java:139)
	at io.netty.channel.DefaultChannelPromise.addListener(DefaultChannelPromise.java:93)
	at io.netty.channel.DefaultChannelPromise.addListener(DefaultChannelPromise.java:28)
	at io.netty.handler.ssl.SslHandler.safeClose(SslHandler.java:1418)
	at io.netty.handler.ssl.SslHandler.closeOutboundAndChannel(SslHandler.java:1243)
	at io.netty.handler.ssl.SslHandler.close(SslHandler.java:459)
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeCloseNow(ChannelHandlerInvokerUtil.java:133)
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeClose(DefaultChannelHandlerInvoker.java:277)
	at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:226)
	at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:185)
	at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.handshakeFailure(ApplicationProtocolNegotiationHandler.java:124)
	at com.relayrides.pushy.apns.ApnsClient$1$1.handshakeFailure(ApnsClient.java:411)
	at io.netty.handler.ssl.ApplicationProtocolNegotiationHandler.userEventTriggered(ApplicationProtocolNegotiationHandler.java:102)
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeUserEventTriggeredNow(ChannelHandlerInvokerUtil.java:75)
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeUserEventTriggered(DefaultChannelHandlerInvoker.java:136)
	at io.netty.channel.AbstractChannelHandlerContext.invokeUserEventTriggered(AbstractChannelHandlerContext.java:350)
	at io.netty.channel.AbstractChannelHandlerContext.fireUserEventTriggered(AbstractChannelHandlerContext.java:138)
	at io.netty.handler.ssl.SslUtils.notifyHandshakeFailure(SslUtils.java:121)
	at io.netty.handler.ssl.SslHandler.notifyHandshakeFailure(SslHandler.java:1222)
	at io.netty.handler.ssl.SslHandler.access$700(SslHandler.java:166)
	at io.netty.handler.ssl.SslHandler$5.run(SslHandler.java:1365)
	at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:339)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:356)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
	at java.lang.Thread.run(Thread.java:745)
Tue Mar 29 20:53:28 PDT 2016: Failed to send apple push notification: com.relayrides.pushy.apns.ClientNotConnectedException
```

**Comment by jchambers** [comment]
With thanks to @ylgrgyq, I think this should be addressed by #261; I can't promise that your connection problems will go away, but at least we should know why connections are failing sometimes. If it seems like something else is going wrong here, please post a comment and we'll investigate.

Cheers!

**Comment by starikoff** [comment]
I have the same problem with 0.7.3. 

Are there any workarounds possible? Should I try connecting several times? This happened in the environment that I don't control, so I'll have to account for such situations in the code... Also, it happened on the first deploy. Deployers refused to proceed further until I do something about it and claim the problem to be resolved (: It's not impossible (though very unlikely) that it's some access control problem on the server.

The code, when stripped off of my constants, is

            apnsClient = new ApnsClient<>(
                    MyApnsClient.class.getResourceAsStream("MyCertificate.p12"),
                    "password");
            Future<Void> connectFuture = apnsClient.connect(
                    ApnsClient.DEVELOPMENT_APNS_HOST,
                    ApnsClient.DEFAULT_APNS_PORT);
            connectFuture.get(1000, TimeUnit.MILLISECONDS); // here the exception happens

The stacktrace is:

```
java.lang.IllegalStateException: Channel closed before HTTP/2 preface completed.
 at com.relayrides.pushy.apns.ApnsClient$2.operationComplete(ApnsClient.java:712)
 at com.relayrides.pushy.apns.ApnsClient$2.operationComplete(ApnsClient.java:704)
 at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:518)
 at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:492)
 at io.netty.util.concurrent.DefaultPromise.notifyListenersWithStackOverFlowProtection(DefaultPromise.java:431)
 at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
 at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:108)
 at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
 at io.netty.channel.AbstractChannel$CloseFuture.setClosed(AbstractChannel.java:1049)
 at io.netty.channel.AbstractChannel$AbstractUnsafe.doClose0(AbstractChannel.java:674)
 at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:652)
 at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:596)
 at io.netty.channel.DefaultChannelPipeline$HeadContext.close(DefaultChannelPipeline.java:1256)
 at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:610)
 at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:594)
 at io.netty.channel.ChannelOutboundHandlerAdapter.close(ChannelOutboundHandlerAdapter.java:71)
 at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:610)
 at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:594)
 at io.netty.handler.ssl.SslHandler.closeOutboundAndChannel(SslHandler.java:1250)
 at io.netty.handler.ssl.SslHandler.close(SslHandler.java:459)
 at io.netty.channel.AbstractChannelHandlerContext.invokeClose(AbstractChannelHandlerContext.java:610)
 at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:594)
 at io.netty.channel.AbstractChannelHandlerContext.close(AbstractChannelHandlerContext.java:451)
 at io.netty.channel.DefaultChannelPipeline.close(DefaultChannelPipeline.java:943)
 at io.netty.channel.AbstractChannel.close(AbstractChannel.java:227)
 at io.netty.channel.ChannelFutureListener$2.operationComplete(ChannelFutureListener.java:56)
 at io.netty.channel.ChannelFutureListener$2.operationComplete(ChannelFutureListener.java:52)
 at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:518)
 at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:511)
 at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:490)
 at io.netty.util.concurrent.DefaultPromise.notifyListenersWithStackOverFlowProtection(DefaultPromise.java:431)
 at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
 at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:126)
 at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:325)
 at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:341)
 at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:580)
 at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:504)
 at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:418)
 at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:390)
 at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
 at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145)
 at java.lang.Thread.run(Thread.java:745)
```

If more information is necessary, I'll be happy to provide it.
Thank you.
